src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -250,231 +220,94 @@ public void AddPeer(ITxPoolPeer peer)
         {
-            if (tx.Hash is null)
-            {
-                throw new ArgumentException($"{nameof(tx.Hash)} not set on {nameof(Transaction)}");
-            }
-            
+            Metrics.PendingTransactionsReceived++;
+
+            // assign a sequence number to transaction so we can order them by arrival times when
+            // gas prices are exactly the same
             tx.PoolIndex = Interlocked.Increment(ref _txIndex);
 
             NewDiscovered?.Invoke(this, new TxEventArgs(tx));
 
             bool managedNonce = (handlingOptions & TxHandlingOptions.ManagedNonce) == TxHandlingOptions.ManagedNonce;
-            bool isPersistentBroadcast = (handlingOptions & TxHandlingOptions.PersistentBroadcast) ==
-                                         TxHandlingOptions.PersistentBroadcast;
-            bool isReorg = (handlingOptions & TxHandlingOptions.Reorganisation) == TxHandlingOptions.Reorganisation;
-
-            if (_logger.IsTrace)
-                _logger.Trace(
-                    $"Adding transaction {tx.ToString("  ")} - managed nonce: {managedNonce} | persistent broadcast {isPersistentBroadcast}");
-
-            return FilterTransaction(tx, managedNonce, isReorg) ?? AddCore(tx, isPersistentBroadcast, isReorg);
-        }
-        
-        protected virtual AddTxResult? FilterTransaction(Transaction tx, in bool managedNonce, in bool isReorg)
-        {
-            Metrics.PendingTransactionsReceived++;
+            bool startBroadcast = (handlingOptions & TxHandlingOptions.PersistentBroadcast) == TxHandlingOptions.PersistentBroadcast;
 
-            if (tx.Hash is null)
-            {
-                Metrics.PendingTransactionsDiscarded++;
-                return AddTxResult.Invalid;
-            }
+            if (_logger.IsTrace) _logger.Trace(
+                $"Adding transaction {tx.ToString("  ")} - managed nonce: {managedNonce} | persistent broadcast {startBroadcast}");
 
-            if (!isReorg && _hashCache.Get(tx.Hash))
-            {
-                Metrics.PendingTransactionsKnown++;
-                return AddTxResult.AlreadyKnown;
-            }
-            
-            IReleaseSpec spec = _specProvider.GetSpec();
-            
-            if (!_validator.IsWellFormed(tx, spec))
-            {
-                // It may happen that other nodes send us transactions that were signed for another chain or don't have enough gas.
-                Metrics.PendingTransactionsDiscarded++;
-                if (_logger.IsTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, invalid transaction.");
-                return AddTxResult.Invalid;
-            }
-            
-            long gasLimit = Math.Min(BlockGasLimit ?? long.MaxValue, _txPoolConfig.GasLimit ?? long.MaxValue);
-            if (tx.GasLimit > gasLimit)
+            foreach (IIncomingTxFilter incomingTxFilter in _filterPipeline)
```

Comment:
Minor: hot path - use for to reduce mem allocations a bit?

Reply:
true, worth changing although for now it will get buried in other allocations -> I was planning to un dotMemory profiling on it -> I would keep dotmemory as a separate task to deliver for txpool